cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -455,66 +477,65 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
+                        if (currentKafkaRebalance != null) {
+                            // Check resource is in the right state as previous execution might have set the status and completed the future
+                            // Safety check as timer might be called again (from a delayed timer firing)
+                            if (KafkaRebalanceState.PendingProposal.equals(state(currentKafkaRebalance))) {
+                                if (KafkaRebalanceAnnotation.stop.equals(rebalanceAnnotation(currentKafkaRebalance))) {
                                     log.debug("{}: Stopping current Cruise Control proposal request timer", reconciliation);
                                     vertx.cancelTimer(t);
                                     p.complete(buildRebalanceStatus(null, KafkaRebalanceState.Stopped));
                                 } else {
                                     requestRebalance(reconciliation, host, apiClient, true, rebalanceOptionsBuilder,
-                                            freshKafkaRebalance.getStatus().getSessionId()).onComplete(rebalanceResult -> {
-                                                if (rebalanceResult.succeeded()) {
-                                                    // If the returned status has an optimization result then the rebalance proposal
-                                                    // is ready, so stop the polling
-                                                    if (rebalanceResult.result().getOptimizationResult() != null &&
-                                                            !rebalanceResult.result().getOptimizationResult().isEmpty()) {
-                                                        vertx.cancelTimer(t);
-                                                        log.debug("{}: Optimization proposal ready", reconciliation);
-                                                        p.complete(rebalanceResult.result());
-                                                    } else {
-                                                        log.debug("{}: Waiting for optimization proposal to be ready", reconciliation);
-                                                    }
-                                                    // The rebalance proposal is still not ready yet, keep the timer for polling
-                                                } else {
-                                                    log.error("{}: Cruise Control getting rebalance proposal failed", reconciliation, rebalanceResult.cause());
-                                                    vertx.cancelTimer(t);
-                                                    p.fail(rebalanceResult.cause());
-                                                }
-                                            });
+                                            currentKafkaRebalance.getStatus().getSessionId())
+                                        .onFailure(e -> {
```

Comment:
can we put the `onFailure` after the `onSuccess` just for having branch code that should be called most of the times as first during the code reading?